src/python/pants/backend/docker/docker_build.py

```diff
@@ -51,24 +66,51 @@ def dockerfile_path(self) -> str:
         return path.join(self.address.spec_path, self.dockerfile_relpath)
 
     def image_names(
-        self, default_name_template: str, registries: DockerRegistries
+        self,
+        default_name_template: str,
+        registries: DockerRegistries,
+        version_context: FrozenDict[str, DockerVersionContextValue],
     ) -> tuple[str, ...]:
         """This method will always return a non-empty tuple."""
         default_parent = path.basename(path.dirname(self.address.spec_path))
         default_repo = path.basename(self.address.spec_path)
         repo = self.repository.value or default_repo
         name_template = self.name_template.value or default_name_template
-        image_name = name_template.format(
-            name=self.name.value or self.address.target_name,
-            repository=repo,
-            sub_repository="/".join(
-                [default_repo if self.repository.value else default_parent, repo]
-            ),
-        )
-        image_names = tuple(
-            ":".join(s for s in [image_name, tag] if s)
-            for tag in [self.version.value, *(self.tags.value or [])]
-        )
+        try:
+            image_name = name_template.format(
+                name=self.name.value or self.address.target_name,
+                repository=repo,
+                sub_repository="/".join(
+                    [default_repo if self.repository.value else default_parent, repo]
+                ),
+            )
+        except KeyError as e:
+            raise DockerNameTemplateError(
+                f"Invalid image name template: {name_template!r}. Unknown key: {e}.\n\n"
+                f"Use any of 'name', 'repository' or 'sub_repository' in the template string."
+            ) from e
+
+        try:
+            image_names = tuple(
+                ":".join(s for s in [image_name, tag] if s)
+                for tag in [
+                    cast(str, self.version.value).format(**version_context),
+                    *(self.tags.value or []),
+                ]
+            )
+        except (KeyError, ValueError) as e:
+            msg = (
+                "Invalid format string for the `docker_image(version)` field: "
```

Comment:
People might not know what you mean by "`docker_image(version)` field". I'd go with "for the `version` field of the docker_image target at {self.address.spec}`

Reply:
I don't have the address at that point, it is added further up the call stack. So the message becomes something like
"Error in {address}: Invalid format string for the `version` field of the docker_image target."

OK? Otherwise I can pass the address to `image_names()` for the sake of the error message..

Reply:
```
  DockerBuildError: Error in testprojects/src/python/docker:test-example: Invalid format string for the `version` field of the docker_image target: '1.2.5-{typo}'.

The key 'typo' is unknown. Try with one of: baseimage, stage0.
```

Reply:
Doh. Dummy, of course I have the address... 🤦🏽